python/beeai_framework/utils/templates.py

```diff
@@ -30,20 +31,44 @@ class Prompt(TypedDict):
 T = TypeVar("T", bound=BaseModel)
 
 
+class PromptTemplateInput(Generic[T]):
+    def __init__(
+        self,
+        schema: type[T],
+        template: str,
+        functions: dict[str, Callable[[], str]] | None = None,
+        defaults: dict[str, str] | None = None,
+    ) -> None:
+        self.schema = schema
+        self.template = template
+        self.functions = functions
+        self.defaults = defaults or {}
+
+
 class PromptTemplate(Generic[T]):
-    def __init__(self, schema: type[T], template: str, functions: dict[str, Callable[[], str]] | None = None) -> None:
-        self._schema: type[T] = schema
-        self._template: str = template
-        self._functions: dict[str, Callable[[], str]] | None = functions
+    def __init__(self, config: PromptTemplateInput) -> None:
+        self._config = config
 
     def render(self, input: ModelLike[T]) -> str:
-        data = to_model(self._schema, input).model_dump()
+        data = to_model(self._config.schema, input).model_dump()
+
+        if self._config.defaults:
+            for key, value in self._config.defaults.items():
+                if data.get(key) is None:
+                    data.update({key: value})
 
         # Apply function derived data
-        if self._functions:
-            for key in self._functions:
+        if self._config.functions:
+            for key in self._config.functions:
                 if key in data:
                     raise PromptTemplateError(f"Function named '{key}' clashes with input data field!")
-                data[key] = self._functions[key]()
+                data[key] = self._config.functions[key]()
+
+        return chevron.render(template=self._config.template, data=data)
 
-        return chevron.render(template=self._template, data=data)
+    def fork(self, customizer: Callable[[PromptTemplateInput], "PromptTemplate"] | None) -> "PromptTemplate":
+        config = copy.copy(self._config)
+        new_config = customizer(config) if customizer else config
+        if isinstance(new_config, dict):
+            raise ValueError("Return type from customizer must be a config or nothing.")
```

Comment:
This check does not seem correct.
Make the `PromptTemplateInput` to inherit from `BaseModel` and then verify whether the output is an instance of it. In addition you will no longer need to use `copy.copy`.